src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,249 @@
+            case (_, cv) =>
+              ScheduleTrigger(
+                EndRefuelSessionTrigger(
+                  tick,
+                  cv.vehicle.getChargerConnectedTick(),
+                  cv.totalChargingSession.energy,
+                  cv.vehicle
+                ),
+                cv.agent
+              )
+          }
+          vehiclesToCharge.clear()
+          completeTriggers
+        }
+      )
+  }
+
+  private def replanHorizon(tick: Int): (Int, Map[Id[BeamVehicle], Double]) = {
+    val requiredPower = sitePowerManager.getPowerOverPlanningHorizon(vehicles)
+
+    val (bounds, nextTick) = if (gridConnectionEnabled) {
+      powerController.publishPowerOverPlanningHorizon(requiredPower, tick)
+      powerController.obtainPowerPhysicalBounds(tick)
+    } else {
+      powerController.defaultPowerPhysicalBounds(tick)
+    }
+    val requiredEnergyPerVehicle = sitePowerManager.replanHorizonAndGetChargingPlanPerVehicle(bounds, vehicles)
+
+    if (requiredEnergyPerVehicle.nonEmpty)
+      log.info("Required energy per vehicle: {}", requiredEnergyPerVehicle.mkString(","))
+
+    (nextTick, requiredEnergyPerVehicle)
+  }
+
+  private def endRefuelSessionTriggerMaybe(
+    vehicle: BeamVehicle,
+    tick: Int,
+    maxChargeDuration: Long,
+    currentSession: ChargingSession,
+    totalSession: ChargingSession
+  ): Option[EndRefuelSessionTrigger] = {
+    if (vehicle.primaryFuelLevelInJoules >= vehicle.beamVehicleType.primaryFuelCapacityInJoule) {
+      log.debug(
+        "Vehicle {} is fully charged. Scheduling EndRefuelSessionTrigger at {} with {} J delivered",
+        vehicle.id,
+        tick + currentSession.duration.toInt,
+        totalSession.energy
+      )
+      Some(
+        EndRefuelSessionTrigger(
+          tick + currentSession.duration.toInt,
+          vehicle.getChargerConnectedTick(),
+          totalSession.energy,
+          vehicle
+        )
+      )
+    } else if (currentSession.duration < maxChargeDuration) {
+      log.debug(
+        "Vehicle {} is charged by a short time: {}. Scheduling EndRefuelSessionTrigger at {} with {} J delivered",
+        vehicle.id,
+        currentSession.duration,
+        tick + currentSession.duration.toInt,
+        totalSession.energy
+      )
+      Some(
+        EndRefuelSessionTrigger(
+          tick + currentSession.duration.toInt,
+          vehicle.getChargerConnectedTick(),
+          totalSession.energy,
+          vehicle
+        )
+      )
+    } else {
+      log.debug(
+        "Ending refuel cycle for vehicle {}. Provided {} J. during {}",
+        vehicle.id,
+        currentSession.energy,
+        currentSession.duration
+      )
+      None
+    }
+  }
+
+  override def postStop: Unit = {
+    log.info("postStop")
+    if (gridConnectionEnabled) {
+      powerController.close()
+    }
+    super.postStop()
+  }
+}
+
+object ChargingNetworkManager {
+  final case class PlanningTimeOutTrigger(tick: Int) extends Trigger
```

Comment:
Should we call this something more explicit to what is being done - unless I misread something? My suggestion is ReevaluateChargingVehicles(Statuses?) or something like that.

Reply:
This name was taken from the sequence diagram from task description - https://github.com/LBNL-UCB-STI/beam/issues/2845

Reply:
@haitamlaarabi Thoughts on the name change - or the domain follows this structure more?